geode-core/src/test/java/org/apache/geode/internal/cache/BucketAdvisorTest.java

```diff
@@ -150,4 +155,42 @@ public void volunteerForPrimaryIgnoresMissingPrimaryElector() {
     advisorSpy.volunteerForPrimary();
     verify(volunteeringDelegate).volunteerForPrimary();
   }
+
+  @Test
+  public void shadowBucketsDestroyedTrackingShouldWorkCorrectly() {
+    DistributionManager distributionManager = mock(DistributionManager.class);
+    when(distributionManager.getId()).thenReturn(new InternalDistributedMember("localhost", 321));
+
+    Bucket bucket = mock(Bucket.class);
+    when(bucket.isHosting()).thenReturn(true);
+    when(bucket.isPrimary()).thenReturn(false);
+    when(bucket.getDistributionManager()).thenReturn(distributionManager);
+
+    PartitionedRegion partitionedRegion = mock(PartitionedRegion.class);
+    when(partitionedRegion.getRedundantCopies()).thenReturn(0);
+    when(partitionedRegion.getPartitionAttributes()).thenReturn(new PartitionAttributesImpl());
+    RegionAdvisor regionAdvisor = mock(RegionAdvisor.class);
+    when(regionAdvisor.getPartitionedRegion()).thenReturn(partitionedRegion);
+
+    List<String> shadowBuckets = Arrays.asList("/bucket1", "/bucket2", "/bucket3");
+    BucketAdvisor bucketAdvisor = BucketAdvisor.createBucketAdvisor(bucket, regionAdvisor);
+    shadowBuckets.forEach(bucketAdvisor::markShadowBucketAsDestroyed);
+
+    // Return false by default.
+    assertThat(bucketAdvisor.isShadowBucketDestroyed("/bucket")).isFalse();
+
+    // Return correct value when found.
+    bucketAdvisor.markShadowBucketAsDestroyed(shadowBuckets.get(1));
+    assertThat(bucketAdvisor.isShadowBucketDestroyed(shadowBuckets.get(1))).isTrue();
+
+    // Mark all shadow buckets values as destroyed
+    bucketAdvisor.markAllShadowBucketsAsDestroyed();
+    shadowBuckets
+        .forEach(b -> assertThat(assertThat(bucketAdvisor.isShadowBucketDestroyed(b)).isTrue()));
+
+    // Mark all shadow buckets values as non destroyed
+    bucketAdvisor.markAllShadowBucketsAsNonDestroyed();
```

Comment:
maybe break this into two tests, one just for marking?  it might seem redundant so it's up to you

Reply:
Changed!